docs/zkapps/o1js/foreign-fields.mdx

```diff
@@ -108,6 +106,214 @@ let y = SmallField.from(5n); // convert from bigint or number
+
+For example, if `x` has the value 16, it is valid for `x.add(x)` to contain the value 32. The addition is correct modulo 17, but doesn't guarantee a result smaller than 17.
+
+:::note
+
+Unreduced doesn't usually mean that the underlying witness is larger than the modulus. It just means that we have not _proved_ it to be smaller.
+A malicious prover _could_ make it larger, by slightly modifying their local version of o1js and creating a proof with that.
+
+:::
+
+Unreduced fields can be added and subtracted, but not multiplied or divided:
+
+```ts
+z.add(1).sub(x); // works
+
+assert((z as any).mul === undefined); // z.mul() is not defined
+assert((z as any).inv === undefined);
+assert((z as any).div === undefined);
+```
+
+### Almost reduced fields
+
+To do multiplication, you need almost reduced fields. You can convert to them by using `.assertAlmostReduced()`:
+
+```ts
+let zAlmost = z.assertAlmostReduced();
+assert(zAlmost instanceof SmallField.AlmostReduced);
+```
+
+Now you can do multiplication and division:
+
+```ts
+let zz = zAlmost.mul(zAlmost); // zAlmost.mul() is defined
+
+// but .mul() returns an unreduced field again:
+assert(zz instanceof SmallField.Unreduced);
+
+// zAlmost.inv() is defined, and returns an almost reduced field:
+assert(zAlmost.inv() instanceof SmallField.AlmostReduced);
+```
+
+It can be convenient to require almost reduced fields as inputs to your smart contract. To do so, create a class which can also serve as a type, and use its `.provable` property when passing to the state decorator:
+
+```ts
+class AlmostField17 extends Field17.AlmostReduced {}
+
+class MyContract extends SmartContract {
+  @state(AlmostField17.provable) x = State<AlmostField17>();
+
+  @method myMethod(y: AlmostField17) {
+    let x = y.mul(2);
+    this.x.set(x.assertAlmostReduced());
+  }
+}
+```
+
+#### What does almost reduced mean?
+
+The definition of almost reduced is somewhat technical. The main motivation is to guarantee that the way we prove modular multiplication is sound. That is definitely true for field elements `< 2^259`. (Recall that we require the modulus to be `< 2^259`.)
+
+However, we actually prove a stronger condition, which lets us save a few constraints in some places:
+
+`z` is **almost reduced** modulo `f`, if `z >> 176` is smaller or equal than `f >> 176`. (`>>` means a [right shift](https://en.wikipedia.org/wiki/Arithmetic_shift).)
+
+:::note
+
+Example: Assume `x` is a `UInt256` holding the value `2^130`. After computing `z = x.mul(x)`, it is valid for `z` to be `2^260`.
+
+However, by calling `z.assertAlmostReduced()`, we prove that `z` is smaller than `2^259` and safe to use in another multiplication. According to our stronger definition, we even have `z < 2^256`.
+
+:::
+
+Why are we exposing `AlmostReducedField` as a separate type, and don't _always_ prove conditions necessary for multiplication? Because that would take up additional constraints!
```

Comment:
```suggestion
However, by calling `z.assertAlmostReduced()`, you prove that `z` is smaller than `2^259` and is safe to use in another multiplication. According to this stronger definition, you even have `z < 2^256`.

:::

Why is `AlmostReducedField` exposed as a separate type, and why is _always_ prove conditions required for multiplication? Because that would take up additional constraints!
```

Reply:
This is a tough one for me, because "we" is so standard in the mathematical arguments I'm used to writing

```suggestion
However, by calling `z.assertAlmostReduced()`, you prove that `z` is smaller than `2^259` and safe to use in another multiplication. According to the stronger definition, you even have `z < 2^256`.

:::

Why is `AlmostReducedField` exposed as a separate type, instead of _always_ proving conditions necessary for multiplication? Because that would take up additional constraints!
```

Reply:
"necessary condition" is standard math language so I'm going to leave that